src/aggregate/aggregate_exec.c

```diff
@@ -223,20 +223,36 @@ static size_t serializeResult(AREQ *req, RedisModule_Reply *reply, const SearchR
         flags |= (req->reqflags & QEXEC_FORMAT_EXPAND) ? SENDREPLY_FLAG_EXPAND : 0;
 
         unsigned int apiVersion = req->sctx->apiVersion;
-        if (v && v->t == RSValue_Duo) {
-          // Which value to use for duo value
+        if (v && v->t == RSValue_JSON) {
+          // Which value to use for JSON value
           if (!(flags & SENDREPLY_FLAG_EXPAND)) {
             // STRING
             if (apiVersion >= APIVERSION_RETURN_MULTI_CMP_FIRST) {
               // Multi
-              v = RS_DUOVAL_OTHERVAL(*v);
+              if(!RS_JSONVAL_SERIALIZED(*v)) {
+                RedisModuleString *serialized;
+                japi->getJSONFromIter(RS_JSONVAL_ITER(*v), reply->ctx, &serialized);
+                RS_JSONVAL_SERIALIZED(*v) = RS_StealRedisStringVal(serialized);
+                if(RS_JSONVAL_SERIALIZED(*v) && RS_JSONVAL_EXPANDED(*v)) {
+                  // free iterator, as it is no longer needed
+                  japi->freeIter(RS_JSONVAL_ITER(*v));
+                  RS_JSONVAL_ITER(*v) = NULL;
+                }
+                v = RS_JSONVAL_SERIALIZED(*v);
+              }
             } else {
               // Single
-              v = RS_DUOVAL_VAL(*v);
+              v = RS_JSONVAL_FIRST(*v);
             }
           } else {
             // EXPAND
-            v = RS_DUOVAL_OTHER2VAL(*v);
+            RS_JSONVAL_EXPANDED(*v) = japi_ver >= 4 ? jsonIterToValueExpanded(reply->ctx, RS_JSONVAL_ITER(*v)) : RS_NullVal();
+            if(RS_JSONVAL_SERIALIZED(*v) && RS_JSONVAL_EXPANDED(*v)) {
+              // free iterator, as it is no longer needed
+              japi->freeIter(RS_JSONVAL_ITER(*v));
```

Comment:
Why do we care freeing it here? wouldn't it get free later anyway?

Reply:
If the RSValue sits in the Sorting Vector, it will be freed later on - no need to keep the iterator that long.

Reply:
On second thought - **this** may not be thread-safe. Will update.